### PR TITLE
Game: Add minigame elements in SA game

### PR DESCRIPTION
**Pull request by Eola-Z**
## Gameplay Additions:
- Created a new GameTable object that can be used to create puzzles inside the SA game. The basic template is a rectangular grid of buttons that store its own state and can modify the state of itself (as well as other buttons inside the table).
- As a proof of concept, we have created 3 simple puzzles that are based on GameTable:
1. A simple light switch puzzle.
2. A combination lock puzzle
3. A rotating pipe puzzle
- Added InputGameElement class which encapsulates an HTML input element and a button as a Phaser game object. Developers can place and position an input box and button over the and retrieve input from them. The aim is to improve personalization and customization in the SA game by allowing players to name themselves, items, pets, etc. The game can then refer to these names in the scripts, creating a more personalized gaming experience.

## Future work/Improvements:

- Pipeline to return the value in the input box to the parser and store it as a state variable for the player, allowing the engine to use that variable for multiple purposes.
- Create new user interface elements based on the InputGameElement template to expand the capabilities of the game frontend.
- Enable the parser to create instances of puzzles and InputGameElements so that story writers can dictate where to add puzzles and input requests to the SA game.
- Remove redundant constants (since switching from volume buttons to sliders, some constants in the files are no longer needed and can be safely removed).
- Once the parser can create puzzles and InputGameElements, we will update the story writer's guide with information on how they can generate such game objects in the text files.

### How to test

Currently, the game objects cannot be added to the game without being hardcoded using an editor.
However, in any Source Academy scene (Phaser), you should be able to instantiate these game objects and add them to the scene itself. Documentation and instructions on how to construct an instance of these game objects can be found the .ts files.

### Created by
Gupta Raman & Edwin Zheng

### CODE REVIEW

**Comment by lhw-1** [comment]
Good work @eyelessrhyme7 @Eola-Z! Some preliminary comments before we get to the review.

First, can we separate the bug fixes (and if possible, the main features) into separate PRs each? This will help declutter the PR & isolate each feature into a single PR. 

Suggestion: maybe you can leave the main feature within this PR and change title & description accordingly, and open new PRs for the bug (and maybe separate the two main features unless you feel it's good to implement them together).

Also, from the PR description:

> However, in any Source Academy scene (Phaser), you should be able to instantiate these game objects and add them to the scene itself. Documentation and instructions on how to construct an instance of these game objects can be found the .ts files.

Let's try and have documentations / instructions updated within the game developer / storywriter guides instead of the `.ts` files.

Remember to refer to other / past PRs for examples on formatting 👍

**Comment by RichDom2185** [comment]
Can I check what's the status of this PR? What is left to do?

**Comment by gok99** [comment]
Tested https://github.com/source-academy/frontend/commit/8026d59dd9d0c8573462fc5e36be71b5dbaab1e8

I was however unable to get the input element to appear, and it's not quite clear how one might use the input. Some work will also be needed to have it be usable in a txt - could be discussed and implemented separately if it's out of scope of this PR.

File: src/features/game/escape/GameEscapeManager.ts [hunk 1]
```diff
@@ -200,15 +206,12 @@ class GameEscapeManager implements IGameUI {
...
-    const bgmVol = this.bgmVolumeRadioButtons
-      ? parseFloat(this.bgmVolumeRadioButtons.getChosenChoice())
-      : 1;
+    const sfxVol = this.sfxVolumeSlider ? this.sfxVolumeSlider.getValue() : 1;
+    const bgmVol = this.bgmVomlumeSlider ? this.bgmVomlumeSlider.getValue() : 1;
+    const muted = this.muteButton ? this.muteButton.getChoice() : false;
 
     // Save settings
-    const newSettings = { bgmVolume: bgmVol, sfxVolume: sfxVol };
+    const newSettings = { bgmVolume: bgmVol, sfxVolume: sfxVol, isMuted: muted };
```
Comment (gok99) [inline src/features/game/escape/GameEscapeManager.ts:214]: Would be nice muting behavior was as follows:
- Mute button placed below both sliders so it doesn't appear visually tied to just SFX or BGM
- isMuted is simply the state of bgmVolume and sfxVolume being 0 so it wouldn't need a separate field. In the UI, selecting the checkbox will simply set the sliders to 0. Similarly setting both sliders to 0 manually will activate the checkbox

What do you think?

**Review by gok99 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**